Core/Src/main.c

```diff
@@ -1,181 +1,215 @@
+  * License. You may obtain a copy of the License at:
+  *                        opensource.org/licenses/BSD-3-Clause
+  *
+  ******************************************************************************
+  */
+/* USER CODE END Header */
+/* Includes ------------------------------------------------------------------*/
+#include "main.h"
+#include "spi.h"
+#include "tim.h"
+#include "usart.h"
+#include "gpio.h"
+
+/* Private includes ----------------------------------------------------------*/
+/* USER CODE BEGIN Includes */
+
+/* USER CODE END Includes */
+
+/* Private typedef -----------------------------------------------------------*/
+/* USER CODE BEGIN PTD */
+
+/* USER CODE END PTD */
+
+/* Private define ------------------------------------------------------------*/
+/* USER CODE BEGIN PD */
+#define SPI_BUF_SIZE 3 // bytes
+#define SPI_TIMEOUT 64 // ticks
+#define ADC_MAX 0x03ff
+#define TIM_MAX_CNT 64000
+/* USER CODE END PD */
+
+/* Private macro -------------------------------------------------------------*/
+/* USER CODE BEGIN PM */
+
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+
+/* USER CODE BEGIN PV */
+
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+/* USER CODE BEGIN PFP */
+
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+
+/* USER CODE END 0 */
+
+/**
+  * @brief  The application entry point.
+  * @retval int
+  */
+int main(void)
+{
+  /* USER CODE BEGIN 1 */
+
+  /* USER CODE END 1 */
+
+  /* MCU Configuration--------------------------------------------------------*/
+
+  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
+  HAL_Init();
+
+  /* USER CODE BEGIN Init */
+
+  uint8_t txd[SPI_BUF_SIZE] = {0x80, 0x01, 0x00};
+  uint8_t rxd[SPI_BUF_SIZE] = {0};
+  uint32_t adc_out = 0;
+  uint32_t compare_value = 0;
+
+  /* USER CODE END Init */
+
+  /* Configure the system clock */
+  SystemClock_Config();
+
+  /* USER CODE BEGIN SysInit */
+
+  /* USER CODE END SysInit */
+
+  /* Initialize all configured peripherals */
+  MX_GPIO_Init();
+  MX_USART2_UART_Init();
+  MX_SPI1_Init();
+  MX_TIM1_Init();
+  /* USER CODE BEGIN 2 */
+
+  // pull up chip select (PB8)
+  HAL_GPIO_WritePin(GPIOA, GPIO_PIN_8, GPIO_PIN_SET);
+  // start timer 1 channel 1
+  HAL_TIM_Base_Start(&htim1);
+
+  /* USER CODE END 2 */
+
+  /* Infinite loop */
+  /* USER CODE BEGIN WHILE */
+  while (1)
+  {
+	  // pull down chip select
+	  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_RESET);
+	  // SPI transaction with SPI1 (return value ignored)
+	  HAL_SPI_TransmitReceive(&hspi1, txd, rxd, SPI_BUF_SIZE, SPI_TIMEOUT);
+	  // compute ADC value from SPI data
+	  adc_out = 0;
+	  for (int i = 0; i < 10; ++ i)
+	  {
+		  adc_out |= ((*((uint32_t *) rxd)) & (1 >> (14 + i))) << (5 + 2*i);

```

Comment:
this is incredibly unreadable, not sure what is going on... I don't think a forloop is required for this transaction... you can stitch together rxd[1] and rxd[2] (which are the only 2 bytes with useful information) in one line of code using bit manipulations